pkg/core/stateroot/module.go

```diff
@@ -20,9 +24,14 @@ type (
 		bc    blockchainer.Blockchainer
 		log   *zap.Logger
 
+		currentLocal    atomic.Value
 		localHeight     atomic.Uint32
 		validatedHeight atomic.Uint32
-		currentLocal    atomic.Value
+
+		mtx              sync.RWMutex
+		validatorsKeys   keys.PublicKeys
+		validatorsHash   util.Uint160
```

Comment:
I don't think you can cache them this way, you can be at 1M state height and 100K stateroot height, and SV list is going to be different between them.

Reply:
Fixed.
The problem is actually deeper: we won't accept payloads for 100K state because of whitelist for `Extensible` which depends on _chain_ height. Also `ValidBlockEnd` for `Extensible` is state height + MaxValidUntilBlockIncrement (another reason why it won't pass validation). There is also a case where some of the state validators go down permanently and committee changes them then (do we have _any_ way to receive validated state root?).
If done correctly we also need (?) to restore full history of state-validator changes from store from validated height to block height in `Init`, because node can go down and up.
This is not even the full list of problems with current implementation of state root processing in C#, so may be we should postpone this until neo-project/neo#2372 and neo-project/neo#2374 is fixed.

Reply:
I am still hoping for state roots in header, because all problems cease to exist then.